core/types/printing.cc

```diff
@@ -398,6 +409,10 @@ string AppliedType::show(const GlobalState &gs) const {
             }
             return to_string(buf);
         } else {
+            if (this->klass.data(gs)->isSingletonClass(gs)) {
+                // T.class_of(klass)[arg1, arg2] is never valid syntax.
+                return this->klass.show(gs);
+            }
```

Comment:
While this is true that this is not (currently) valid syntax, it does make error messages substantially more understandable when working with complex (generic) code. For example:

```ruby
# typed: true
extend T::Sig

class A
  extend T::Sig
  extend T::Generic

  Elem = type_template
end

T.reveal_type(A) # error: `T.class_of(A)[T.untyped]`
```

To a lot of people, they might not realize that:

- Sorbet treats this like an AppliedType (generic) at all
- the `type_template` has been defaulted to `T.untyped`

Again, would love to see how this comes up in practice so we can know whether there's a better heuristic.

Reply:
`TypePtr::show` is documented to be "what the user typed", and "showWithMoreInfo" is "not necessarily what the user typed, but useful for error messages".

Could we just use `showWithMoreInfo` more often?

Reply:
I feel pretty strongly that no. By far the more common use case is error messages, and I think that the shortest name should be reserved for the most common use case. I added `showWithMoreInfo` as a hack, and it's currently only used in a handful of places (hover, reveal type).

If we wanted to change the names and make it like `show` and `showPrintable` I'd be fine with that.